### PR TITLE
Adding packaging

### PR DESCRIPTION
**Pull request by aanm**

### CODE REVIEW

File: bpf/Makefile [hunk 1]
```diff
@@ -40,4 +40,6 @@ install:
 	cp -r include/* $(DESTDIR)$(LIBDIR)/cilium/include/
 	$(INSTALL) -m 0755 -d $(DESTDIR)$(LIBDIR)/cilium/probes/
 	cp -r probes/* $(DESTDIR)$(LIBDIR)/cilium/probes/
+	find $(DESTDIR)$(LIBDIR)/cilium -type d -exec chmod 755 {} \;
+	find $(DESTDIR)$(LIBDIR)/cilium -type f -exec chmod 644 {} \;
```
Comment (tgraf) [inline bpf/Makefile]: Why is this needed? Are the permissions in the git tree wrong? There are scripts in $(LIBDIR)/cilium which have the execute bit set which would be overwritten by this.

File: cilium/Makefile [hunk 2]
```diff
@@ -18,6 +18,7 @@ $(TARGET): $(SOURCES)
 
 install: all
 	$(INSTALL) -m 0755 $(TARGET) $(DESTDIR)$(BINDIR)
-	cp ../contrib/autocomplete/bash_autocomplete /etc/bash_completion.d/cilium
-	chmod 644 /etc/bash_completion.d/cilium
+	$(INSTALL) -m 0755 -d $(DESTDIR)$(CONFDIR)/bash_completion.d
+	cp ../contrib/autocomplete/bash_autocomplete $(DESTDIR)$(CONFDIR)/bash_completion.d/cilium
+	chmod 644 $(DESTDIR)$(CONFDIR)/bash_completion.d/cilium
```
Comment (tgraf) [inline cilium/Makefile]: Can you use $(INSTALL) here as well?

**Review by tgraf — CHANGES_REQUESTED** [review]

**Comment by aanm** [comment]
@tgraf Fixed your issues. I've continued working on this branch so that we can create a rpm and deb package.

**Comment by aanm** [comment]
Some notes for discussion:
- The omission of the policy-watcher and socket-proxy in systemd.
- The hardcode of `linux-image-4.8.1-040801-generic` (and not linux-image (>=4.8) ) in deb

File: bpf/Makefile [hunk 3]
```diff
@@ -43,3 +41,25 @@ install:
...
+install:
+	$(INSTALL) -m 0644 -t $(DESTDIR)$(LIBDIR)/cilium/ $(BPF:.o=.c)
+	$(INSTALL) -m 0755 -t $(DESTDIR)$(LIBDIR)/cilium/ $(SCRIPTS)
+	$(INSTALL) -m 0644 -t $(DESTDIR)$(LIBDIR)/cilium/lib/ $(LIB)
+	for dir in `find include probes -type d`; do \
+		$(INSTALL) -m 0755 -d $(DESTDIR)$(LIBDIR)/cilium/$$dir; \
+		for f in `find $$dir -maxdepth 1 -type f`; do \
+			$(INSTALL) -m 0644 -t $(DESTDIR)$(LIBDIR)/cilium/$$dir $$f; \
+		done; \
+	done
```
Comment (tgraf) [inline bpf/Makefile:53]: Is the install section a duplicate?

File: contrib/packaging/deb/cfg/debian/control [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+Standards-Version: ${VERSION}
+Homepage: https://github/cilium/cilium
+Vcs-Git: git://github/cilium/cilium.git
+
+Package: cilium
+Architecture: amd64
+Section: net
+Priority: optional
+Depends: docker-engine (>=1.12), clang-3.8, libc6-dev-i386, iproute2 (>=4.6.0),
+ linux-image-4.8.1-040801-generic
```
Comment (tgraf) [inline contrib/packaging/deb/cfg/debian/control:14]: Can refer to a minimal kernel version instead?

File: contrib/upstart/cilium-net-daemon.conf [hunk 5]
```diff
@@ -1,2 +1,2 @@
 env PATH=/usr/local/clang/bin:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/sbin:/sbin:/bin
-exec cilium -D daemon run -d eth1
+exec cilium -D daemon run --ipv4 --lb
```
Comment (tgraf) [inline contrib/upstart/cilium-net-daemon.conf]: The `--lb` here looks wrong. Why is this needed?

File: contrib/systemd/cilium-consul.service [hunk 6]
```diff
@@ -0,0 +1,19 @@
+[Unit]
+Description=cilium-consul
+Documentation=https://github.com/cilium/cilium
+Requires=docker.service
```
Comment (tgraf) [inline contrib/systemd/cilium-consul.service:4]: Long term we should probably have a cilium-consul meta unit which can be implemented by either cilium-consul-docker or cilium-consul-rkt. Not sure what the systemd native way is to do this.

File: bpf/Makefile (hunk 3 above)
Comment (aanm) [reply bpf/Makefile:53]: A duplicate of what? The first for loop is for directories inside `include` and `probes`. The second is for all files inside `include` and `probes`.

File: contrib/packaging/deb/cfg/debian/control (hunk 4 above)
Comment (aanm) [reply contrib/packaging/deb/cfg/debian/control:14]: No, the version is in the name of the kernel package. We can omit it and write in **bold** the minimal kernel version that cilium needs in the installation readme

File: contrib/systemd/cilium-consul.service (hunk 6 above)
Comment (aanm) [reply contrib/systemd/cilium-consul.service:4]: Ah I forgot to write that on the discussion points. I think in long term we should provide a systemd consul configuration (without running in a container) but also allowing the user to provide us its own consul instance IP.

File: contrib/upstart/cilium-net-daemon.conf (hunk 5 above)
Comment (aanm) [reply contrib/upstart/cilium-net-daemon.conf]: Shouldn't the loadbalancer mode W-E be enabled by default? Isn't this the flag?

File: contrib/packaging/deb/cfg/debian/control (hunk 4 above)
Comment (tgraf) [reply contrib/packaging/deb/cfg/debian/control:14]: Shouldn't we be referring to the meta package instead? `linux-image-amd64`?

File: contrib/upstart/cilium-net-daemon.conf (hunk 5 above)
Comment (tgraf) [reply contrib/upstart/cilium-net-daemon.conf]: It shouldn't need a flag, we enable the LB by default right now.

File: contrib/systemd/cilium-consul.service (hunk 6 above)
Comment (tgraf) [reply contrib/systemd/cilium-consul.service:4]: Sounds good

**Comment by tgraf** [comment]
@aanm Regarding omission of socket-proxy and policy-watcher. That's fine.

File: contrib/packaging/deb/cfg/debian/control (hunk 4 above)
Comment (aanm) [reply contrib/packaging/deb/cfg/debian/control:14]: There's no difference. In my installation tests I've been using [this](http://kernel.ubuntu.com/~kernel-ppa/mainline/v4.8.1/) page to install the kernel since it's not available yet in the testing repos.

File: contrib/upstart/cilium-net-daemon.conf (hunk 5 above)
Comment (aanm) [reply contrib/upstart/cilium-net-daemon.conf]: No, right now is false by default.

``` go
cli.BoolFlag{
        Destination: &config.LBMode,
        Name:        "lb",
        Usage:       "Enables load balancer mode where load balancer bpf program is attached to the interface ",
},
```

Should I change it to `cli.BoolTFlag` and remove the `--lb` flag in the `contrib/upstart/cilium-net-daemon.conf`?

Comment (tgraf) [reply contrib/upstart/cilium-net-daemon.conf]: Yes but the LB functionality is enabled regardless of the `LBMode` value. I thought we added `LBMode` to signal the use case of a standalone LB without additional forwarding to local containers, i.e. listening on a `net_device` and then perform LB if configured.

It is completely pointless if a user has to enable `--lb` to use LB functionality on a regular node. I see that the k8s service integration currently depends on `--lb`. That must be changed, see `daemon/daemon.go:120`. Providing `--lb` is in conflict with direct routing mode.

File: contrib/packaging/deb/cfg/debian/control (hunk 4 above)
Comment (tgraf) [reply contrib/packaging/deb/cfg/debian/control:14]: OK, if there is no official package yet then that's fine but we have to change this the minute an official 4.8 package is available as this will prevent anybody from upgrading their kernels.

File: bpf/Makefile (hunk 3 above)
Comment (tgraf) [reply bpf/Makefile:53]: You are duplicating the code of the install target while it looks like most of it is identical.

Comment (aanm) [reply bpf/Makefile:53]: Ah now I understood what you were talking about. The problem was that `$(MAKE) -C go install` was supposed to be run when `PKG_BUILD==""` but not otherwise. I've made a change, is it acceptable like that?

File: contrib/upstart/cilium-net-daemon.conf (hunk 5 above)
Comment (aanm) [reply contrib/upstart/cilium-net-daemon.conf]: So should k8s service integration assume the cilium is always running in LB mode? And what about `daemon/daemon.go:223`?

File: bpf/Makefile (hunk 3 above)
Comment (tgraf) [reply bpf/Makefile:53]: Yes, looks good.

File: contrib/upstart/cilium-net-daemon.conf (hunk 5 above)
Comment (tgraf) [reply contrib/upstart/cilium-net-daemon.conf]: Yes, if k8s is enabled, services should always be handled.

The maps can be created unconditionally. As for whether to clear them at startup, we can make use of the `--keep-config` flag to opt out of this behavior.

Comment (aanm) [reply contrib/upstart/cilium-net-daemon.conf]: I've removed the `--lb`. I'll make the k8s changes on a different PR.

**Review by tgraf — APPROVED** [review]
